pyomo/contrib/observer/model_observer.py

```diff
@@ -0,0 +1,1294 @@
+            self._check_to_remove_params(self._params_referenced_by_var[v])
+            self._params_referenced_by_var.pop(v)
+            if any(self._referenced_variables[v]):
+                raise ValueError(
+                    f'Cannot remove variable {v.name} - it is still being used by constraints/objectives'
+                )
+            self._referenced_variables.pop(v)
+            self._vars.pop(v)
+
+    def _remove_parameters(self, params: Collection[ParamData]):
+        for p in params:
+            if p not in self._referenced_params:
+                raise ValueError(
+                    f'Cannot remove parameter {p.name} - it has not been added'
+                )
+            self._updates.params_to_update[p] |= Reason.removed
+            if any(self._referenced_params[p]):
+                raise ValueError(
+                    f'Cannot remove parameter {p.name} - it is still being used by constraints/objectives'
+                )
+            self._referenced_params.pop(p)
+            self._params.pop(p)
+
+    def _update_var_bounds(self, v: VarData):
+        ref_params = ComponentSet()
+        for bnd in (v._lb, v._ub):
+            if bnd is None or type(bnd) in native_numeric_types:
+                continue
+            (named_exprs, _vars, parameters, external_functions) = (
+                collect_components_from_expr(bnd)
+            )
+            if _vars:
+                raise NotImplementedError(
+                    'ModelChangeDetector does not support variables in the bounds of other variables'
+                )
+            if named_exprs:
+                raise NotImplementedError(
+                    'ModelChangeDetector does not support Expressions in the bounds of other variables'
+                )
+            if external_functions:
+                raise NotImplementedError(
+                    'ModelChangeDetector does not support external functions in the bounds of other variables'
+                )
+            ref_params.update(parameters)
+
+        _ref_params = self._params_referenced_by_var[v]
+        new_params = ref_params - _ref_params
+        old_params = _ref_params - ref_params
+
+        self._params_referenced_by_var[v] = ref_params
+
+        if new_params:
+            self._check_for_new_params(new_params)
+
+        for p in new_params:
+            self._referenced_params[p][3].add(v)
+        for p in old_params:
+            self._referenced_params[p][3].remove(v)
+
+        if old_params:
+            self._check_to_remove_params(old_params)
+
+    def _update_variables(self, variables: Optional[Collection[VarData]] = None):
+        if variables is None:
+            variables = self._vars
+        for v in variables:
+            _lb, _ub, _fixed, _domain_interval, _value = self._vars[v]
+            lb, ub, fixed, domain_interval, value = (
+                v._lb,
+                v._ub,
+                v.fixed,
+                v.domain.get_interval(),
```

Comment:
This is OK for now, but:
- will break down as soon as we have solver interfaces that support anything other than simple bounded domains (e.g., this won't work for semi-continuous domains)
- `get_interval()` can be inefficient for custom sets

The long-term solution is to probably maintain a list of known domains and perform change detection on them (much like we do for Params).  The short-term solution is probably to document this issue in a comment here.